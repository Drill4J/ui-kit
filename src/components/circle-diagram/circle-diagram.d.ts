/// <reference types="react" />
interface CircleDiagramValues {
    coverage: number;
    diameter: number;
    strokeWidth: number;
}
interface Props {
    className?: string;
    values: CircleDiagramValues;
    children?: React.ReactNode;
}
export declare const CircleDiagram: ({ className, values: { coverage, diameter, strokeWidth }, children, }: Props) => JSX.Element;
export {};
