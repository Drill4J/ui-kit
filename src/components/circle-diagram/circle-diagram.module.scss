@import '../../style-guide/index';

.circle-diagram {
  position: relative;
  display: inline-block;

  circle {
    fill: none;
    stroke: map-get($monochrome, medium-tint);
  }

  circle[type='coverage'] {
    fill: none;
    stroke: map-get($green, default);
    transition: stroke-dashoffset 1000ms ease-in-out;
    opacity: 40%;
  }

  &__children-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    display: block;
  }
}
