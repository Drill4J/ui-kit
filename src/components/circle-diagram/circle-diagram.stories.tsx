import React from 'react';
import { storiesOf } from '@storybook/react';

import { CircleDiagram } from './circle-diagram';

storiesOf('CircleDiagram', module)
  .add('CircleDiagram', () => {
    const [coverage, setCoverage] = React.useState(0);
    const handleChange = ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
      setCoverage(+value);
    };

    React.useEffect(() => {
      const interval = setInterval(() => {
        setCoverage(state => (state < 100 ? state + 10 : 0));
      }, 1000);

      return () => {
        clearInterval(interval);
      };
    }, []);
    return (
      <>
        <div>
          <CircleDiagram values={{
            coverage,
            diameter: 96,
            strokeWidth: 20,
          }}
          >
            <div style={{ width: '50px', fontSize: '11px' }}>
              <div>Counter</div>
              <div style={{ textAlign: 'center' }}>{coverage}</div>
            </div>
          </CircleDiagram>
          <input type="range" min="0" max="100" value={coverage} onChange={handleChange} />
        </div>
        <CircleDiagram values={{
          coverage,
          diameter: 400,
          strokeWidth: 10,
        }}
        >
          <span>Some text</span>
        </CircleDiagram>
      </>
    );
  });
