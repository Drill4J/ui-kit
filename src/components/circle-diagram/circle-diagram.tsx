import { BEM } from '@redneckz/react-bem-helper';

import styles from './circle-diagram.module.scss';

interface CircleDiagramValues {
  coverage: number;
  diameter: number;
  strokeWidth: number;
}

interface Props {
  className?: string;
  values: CircleDiagramValues;
  children?: React.ReactNode;
}

const circleDiagram = BEM(styles);

export const CircleDiagram = circleDiagram(({
  className, values: { coverage, diameter, strokeWidth }, children,
}: Props) => {
  const radius = (diameter - strokeWidth) / 2;
  const circumference = 2 * Math.PI * radius;
  const center = diameter / 2;

  const coverageOffset = ((100 - coverage) / 100) * circumference;

  return (
    <div className={className}>
      <svg
        width={diameter}
        height={diameter}
        transform="rotate(-90)"
      >
        <circle
          cx={center}
          cy={center}
          r={radius}
          strokeWidth={strokeWidth}
        />
        <circle
          type="coverage"
          cx={center}
          cy={center}
          r={radius}
          strokeWidth={strokeWidth}
          strokeDasharray={circumference}
          strokeDashoffset={coverageOffset}
        />
      </svg>
      <ChildrenWrapper>{children}</ChildrenWrapper>
    </div>
  );
});

const ChildrenWrapper = circleDiagram.childrenWrapper('div');
