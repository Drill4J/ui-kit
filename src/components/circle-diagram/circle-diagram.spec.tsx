import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { CircleDiagram } from './circle-diagram';

describe('CircleDiagram', () => {
  const mockValues = {
    coverage: 50,
    diameter: 400,
    strokeWidth: 10,
  };

  it('should match snapshot circle-diagram', () => {
    const tree = renderer.create(
      <CircleDiagram values={mockValues} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should recalculate the offset value when changing coverage', () => {
    const wrapper = mount(
      <CircleDiagram values={mockValues} />,
    );
    wrapper.setProps({ values: { ...mockValues, coverage: 100 } });
    expect(wrapper.find('circle[type="coverage"]').getElement().props.strokeDashoffset).toEqual(0);
  });
});
