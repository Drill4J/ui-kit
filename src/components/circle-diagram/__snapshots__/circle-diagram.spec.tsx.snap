// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleDiagram should match snapshot circle-diagram 1`] = `
<div
  className="circle-diagram"
>
  <svg
    height={400}
    transform="rotate(-90)"
    width={400}
  >
    <circle
      cx={200}
      cy={200}
      r={195}
      strokeWidth={10}
    />
    <circle
      cx={200}
      cy={200}
      r={195}
      strokeDasharray={1225.2211349000193}
      strokeDashoffset={612.6105674500096}
      strokeWidth={10}
      type="coverage"
    />
  </svg>
  <div
    className="circle-diagram__children-wrapper"
    data-modifiers=""
  />
</div>
`;
